File: typecast-java/src/main/java/com/neosapience/models/TTSWithTimestampsResponse.java
package com.neosapience.models;

import com.google.gson.annotations.SerializedName;
import com.neosapience.internal.CaptioningHelpers;
import com.neosapience.internal.CaptioningHelpers.Cue;
import com.neosapience.internal.CaptioningHelpers.Segment;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Base64;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Response from POST /v1/text-to-speech/with-timestamps.
 *
 * <p>Contains the synthesized audio (base64-encoded) together with word and/or
 * character-level alignment segments. Use {@link #toSrt()} / {@link #toVtt()} to
 * produce caption files.</p>
 */
public class TTSWithTimestampsResponse {

    private final String audio;

    @SerializedName("audio_format")
    private final String audioFormat;

    @SerializedName("audio_duration")
    private final double audioDuration;

    private final List<AlignmentSegmentWord> words;
    private final List<AlignmentSegmentCharacter> characters;

    /**
     * Creates a TTSWithTimestampsResponse.
     *
     * @param audio         base64-encoded audio data
     * @param audioFormat   format string (e.g., "wav" or "mp3")
     * @param audioDuration total audio duration in seconds
     * @param words         word-level alignment segments (may be null)
     * @param characters    character-level alignment segments (may be null)
     */
    public TTSWithTimestampsResponse(String audio, String audioFormat, double audioDuration,
                                     List<AlignmentSegmentWord> words,
                                     List<AlignmentSegmentCharacter> characters) {
        this.audio = audio;
        this.audioFormat = audioFormat;
        this.audioDuration = audioDuration;
        this.words = words;
        this.characters = characters;
    }

    /** Returns the base64-encoded audio field. */
    public String getAudio() {
        return audio;
    }

    /** Returns the audio format string. */
    public String getAudioFormat() {
        return audioFormat;
    }

    /** Returns the total audio duration in seconds. */
    public double getAudioDuration() {
        return audioDuration;
    }

    /** Returns the word-level alignment segments, or null if not present. */
    public List<AlignmentSegmentWord> getWords() {
        return words;
    }

    /** Returns the character-level alignment segments, or null if not present. */
    public List<AlignmentSegmentCharacter> getCharacters() {
        return characters;
    }

    /**
     * Decodes and returns the audio data.
     *
     * @return decoded audio bytes
     */
    public byte[] audioBytes() {
        return Base64.getDecoder().decode(audio);
    }

    /**
     * Writes the decoded audio bytes to the given path.
     *
     * @param path target file path
     * @throws IOException if an I/O error occurs
     */
    public void saveAudio(Path path) throws IOException {
        Files.write(path, audioBytes());
    }

    /**
     * Returns SRT-formatted caption text.
     *
     * <p>Cue indices start at 1. Timestamps use {@code HH:MM:SS,mmm} format.
     * Lines are separated by LF.</p>
     *
     * @return SRT content as a string
     * @throws IllegalStateException if there are no usable alignment segments
     */
    public String toSrt() {
        return formatCaptions(true);
    }

    /**
     * Returns WebVTT-formatted caption text.
     *
     * <p>Starts with {@code WEBVTT\n\n}. Timestamps use {@code HH:MM:SS.mmm} format.
     * Lines are separated by LF.</p>
     *
     * @return VTT content as a string
     * @throws IllegalStateException if there are no usable alignment segments
     */
    public String toVtt() {
        return formatCaptions(false);
    }

    // -----------------------------------------------------------------------
    // Private helpers
    // -----------------------------------------------------------------------

    private String formatCaptions(boolean srt) {
        List<Segment> segs;
        boolean wordMode;
        if (words != null && words.size() >= 2) {
            segs = words.stream()
                    .map(w -> new Segment(w.getText(), w.getStart(), w.getEnd()))
                    .collect(Collectors.toList());
            wordMode = true;
        } else if (characters != null && characters.size() >= 1) {
            segs = characters.stream()
                    .map(c -> new Segment(c.getText(), c.getStart(), c.getEnd()))
                    .collect(Collectors.toList());
            wordMode = false;
        } else if (words != null && words.size() == 1) {
            segs = words.stream()
                    .map(w -> new Segment(w.getText(), w.getStart(), w.getEnd()))
                    .collect(Collectors.toList());
            wordMode = true;
        } else {
            throw new IllegalStateException("no alignment segments to caption from");
        }

        List<Cue> cues = CaptioningHelpers.groupIntoCues(segs, wordMode);
        if (cues.isEmpty()) {
            throw new IllegalStateException("no alignment segments to caption from");
        }

        StringBuilder sb = new StringBuilder();
        if (!srt) {
            sb.append("WEBVTT\n\n");
        }
        for (int i = 0; i < cues.size(); i++) {
            Cue c = cues.get(i);
            if (srt) {
                sb.append(i + 1).append('\n');
            }
            sb.append(srt ? CaptioningHelpers.formatSrtTime(c.start)
                          : CaptioningHelpers.formatVttTime(c.start))
              .append(" --> ")
              .append(srt ? CaptioningHelpers.formatSrtTime(c.end)
                          : CaptioningHelpers.formatVttTime(c.end))
              .append('\n');
            sb.append(c.text).append('\n').append('\n');
        }
        return sb.toString();
    }
}

File: typecast-php/src/Models/TTSWithTimestampsResponse.php
<?php

declare(strict_types=1);

namespace Neosapience\Typecast\Models;

use Neosapience\Typecast\Internal\CaptioningHelpers;

/**
 * Response from POST /v1/text-to-speech/with-timestamps.
 *
 * Provides helpers to decode the audio payload and generate SRT/VTT captions
 * from the embedded word- or character-level alignment segments.
 */
final class TTSWithTimestampsResponse
{
    /**
     * @param list<AlignmentSegmentWord>|null      $words
     * @param list<AlignmentSegmentCharacter>|null $characters
     */
    public function __construct(
        public readonly string $audio,
        public readonly string $audioFormat,
        public readonly float $audioDuration,
        public readonly ?array $words,
        public readonly ?array $characters,
    ) {
    }

    /**
     * Build a response object from a decoded JSON array.
     *
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        foreach (['audio', 'audio_format', 'audio_duration'] as $required) {
            if (!array_key_exists($required, $data)) {
                throw new \InvalidArgumentException("Missing required field: {$required}");
            }
        }

        $words = null;
        if (isset($data['words']) && is_array($data['words'])) {
            $words = array_map(
                static fn(array $w) => new AlignmentSegmentWord(
                    (string) $w['text'],
                    (float) $w['start'],
                    (float) $w['end'],
                ),
                $data['words'],
            );
        }

        $chars = null;
        if (isset($data['characters']) && is_array($data['characters'])) {
            $chars = array_map(
                static fn(array $c) => new AlignmentSegmentCharacter(
                    (string) $c['text'],
                    (float) $c['start'],
                    (float) $c['end'],
                ),
                $data['characters'],
            );
        }

        return new self(
            audio: (string) $data['audio'],
            audioFormat: (string) $data['audio_format'],
            audioDuration: (float) $data['audio_duration'],
            words: $words,
            characters: $chars,
        );
    }

    /**
     * Decode the base64-encoded audio payload and return raw bytes.
     *
     * @throws \RuntimeException on invalid base64 data
     */
    public function audioBytes(): string
    {
        $decoded = base64_decode($this->audio, true);
        if ($decoded === false) {
            throw new \RuntimeException('Invalid base64 audio data');
        }
        return $decoded;
    }

    /**
     * Decode and write the audio bytes to a file.
     *
     * @throws \RuntimeException if the write fails
     */
    public function saveAudio(string $path): void
    {
        if (file_put_contents($path, $this->audioBytes()) === false) {
            throw new \RuntimeException("Failed to write audio to {$path}");
        }
    }

    /**
     * Generate an SRT subtitle string from the alignment segments.
     *
     * @throws \RuntimeException when no alignment segments are available
     */
    public function toSrt(): string
    {
        return $this->formatCaptions(srt: true);
    }

    /**
     * Generate a WebVTT subtitle string from the alignment segments.
     *
     * @throws \RuntimeException when no alignment segments are available
     */
    public function toVtt(): string
    {
        return $this->formatCaptions(srt: false);
    }

    /**
     * Build the SRT or VTT text from cues produced by the captioning helpers.
     */
    private function formatCaptions(bool $srt): string
    {
        $picked = $this->pickSegments();
        $cues = CaptioningHelpers::groupIntoCues($picked['segments'], $picked['wordMode']);

        if (count($cues) === 0) {
            throw new \RuntimeException('no alignment segments to caption from');
        }

        $out = '';
        if (!$srt) {
            $out .= "WEBVTT\n\n";
        }

        foreach ($cues as $i => $c) {
            if ($srt) {
                $out .= ($i + 1) . "\n";
            }
            $tStart = $srt
                ? CaptioningHelpers::formatSrtTime($c['start'])
                : CaptioningHelpers::formatVttTime($c['start']);
            $tEnd = $srt
                ? CaptioningHelpers::formatSrtTime($c['end'])
                : CaptioningHelpers::formatVttTime($c['end']);
            $out .= "{$tStart} --> {$tEnd}\n";
            $out .= $c['text'] . "\n\n";
        }

        return $out;
    }

    /**
     * Pick the best available segment list and return it together with a flag
     * indicating whether it is word-mode (true) or character-mode (false).
     *
     * Priority: words (≥2) > characters (≥1) > words (=1)
     *
     * @return array{segments: list<array{text: string, start: float, end: float}>, wordMode: bool}
     * @throws \RuntimeException when neither words nor characters are present
     */
    private function pickSegments(): array
    {
        $toMap = static fn(array $segs): array => array_map(
            static fn($s) => ['text' => $s->text, 'start' => $s->start, 'end' => $s->end],
            $segs,
        );

        $w = $this->words;
        $c = $this->characters;

        if ($w !== null && count($w) >= 2) {
            return ['segments' => $toMap($w), 'wordMode' => true];
        }
        if ($c !== null && count($c) >= 1) {
            return ['segments' => $toMap($c), 'wordMode' => false];
        }
        if ($w !== null && count($w) === 1) {
            return ['segments' => $toMap($w), 'wordMode' => true];
        }

        throw new \RuntimeException('no alignment segments to caption from');
    }
}
